nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py

```diff
@@ -0,0 +1,33 @@
+from nmdc_schema.migrators.migrator_base import MigratorBase
+
+
+class Migrator(MigratorBase):
+    r"""Migrates a database between two schemas."""
+
+    _from_version = "10.3.0"
+    _to_version = "10.4.0"
+
+    def upgrade(self) -> None:
+        r"""Migrates the database from conforming to the original schema, to conforming to the new schema."""
+
+        self.adapter.process_each_document(
+            "nom_analysis_activity_set", [self.update_nom_analysis_activity_id]
+        )
+
+    def update_nom_analysis_activity_id(self, nom_analysis_activity: dict) -> dict:
+        r"""
+        Updates a nom_analysis_activity's id to contain a version number (.1),
+        if it has no version number.
+
+        >>> m = Migrator()
+        >>> m.update_nom_analysis_activity_id({'id': '1'}) # test: adds version number to id
+        {'id': '1.1'}
+        """
+        self.logger.info(f"Processing NomAnalysisActivity: {nom_analysis_activity['id']}")
+
+        
+        # If this data object's `data_object_type` consists of the invalid value, replace it with the valid value.
+        if "." not in nom_analysis_activity['id']:
```

Comment:
I am wondering whether this is a sufficient check. In other words, I am wondering whether it's possible that a "non-versioned" ID still contains a period for an unrelated reason.

Reply:
Here's a regex pattern you can use to determine whether the ID "ends with a dot followed by one or more numeric digits":

```py
pattern = re.compile(r"\.[0-9]+$")  # ends with dot and number(s)
```